draft-ietf-mls-protocol.md

```diff
@@ -1192,40 +1192,36 @@ The details of application key derivation are described in the
-provide some public information about a user.  ClientInitKey
-messages provide information about a client that any existing
-member can use to add this client to the group asynchronously.
-
-A ClientInitKey object specifies what ciphersuites a client supports,
-as well as providing public keys that the client can use for key
-derivation and signing.  The client's identity key is intended to be
-stable throughout the lifetime of the group; there is no mechanism to
-change it.  Init keys are intended to be used a very limited number of
-times, potentially once. (see {{init-key-reuse}}).  ClientInitKeys
-also contain an identifier chosen by the client, which the client
-MUST assure uniquely identifies a given ClientInitKey object among the
-set of ClientInitKeys created by this client.
-
-The init\_keys array MUST have the same length as the cipher\_suites
-array, and each entry in the init\_keys array MUST be a public key
-for the asymmetric encryption scheme defined in the cipher\_suites array
-and used in the HPKE construction for TreeKEM.
-
-The whole structure is signed using the client's identity key.
-A ClientInitKey object with an invalid signature field MUST be
-considered malformed.  The input to the signature computation
-comprises all of the fields except for the signature field.
+In order to facilitate asynchronous addition of clients to a group, it is
+possible to pre-publish initialization keys that provide some public information
+about a user.  ClientInitKey messages provide information about a client that
+any existing member can use to add this client to the group asynchronously.
+
+A ClientInitKey object specifies a ciphersuite that the client supports, as well
+as providing a public key that others can use for key agreement. The client's
+identity key is intended to be stable throughout the lifetime of the group;
+there is no mechanism to change it.  Init keys are intended to be used a very
+limited number of times, ideally only once. (See {{init-key-reuse}}). Clients
+MAY generate and publish multiple ClientInitKey objects to support multiple
+ciphersuites, or to reduce the likelihood of init key reuse. ClientInitKeys
+contain an identifier chosen by the client, which the client MUST assure
+uniquely identifies a given ClientInitKey object among the set of ClientInitKeys
+created by this client.
+
+The value for init\_key MUST be a public key for the asymmetric encryption
+scheme defined by cipher\_suite. The whole structure is signed using the
+client's identity key. A ClientInitKey object with an invalid signature field
+MUST be considered malformed.  The input to the signature computation comprises
+all of the fields except for the signature field.
 
 ~~~~~
 uint8 ProtocolVersion;
 
 struct {
     opaque client_init_key_id<0..255>;
-    ProtocolVersion supported_versions<0..255>;
-    CipherSuite cipher_suites<0..255>;
-    HPKEPublicKey init_keys<1..2^16-1>;
+    ProtocolVersion supported_version;
```

Comment:
Please move this to the first slot in the struct.  That way we can have a very simple protocol invariant -- the first two octets of the InitKey tell you the format of the rest.  (We should write down these invariants, but that's for a different PR.)

Reply:
Protocol version is only one octet, is there something else you want here?

Reply:
Misstatement on my part, sorry.